lib/membrane/exceptions.ex

```diff
@@ -55,81 +51,75 @@ defmodule Membrane.ActionError do
     "Cannot invoke this action in #{playback_state} state."
   end
 
   defp format_reason(:actions_after_redemand) do
     "Redemand action has to be last in actions list."
   end
 
-  defp format_reason({:invalid_buffer, buffer}) do
-    "Invalid buffer: #{inspect(buffer, pretty: true)}"
-  end
-
-  defp format_reason({:unknown_pad, pad}) do
-    "The pad #{inspect(pad)} does not exist"
-  end
-
-  defp format_reason({:invalid_pad_dir, pad, dir}) do
-    "Cannot invoke this action on pad #{inspect(pad)} as it has #{inspect(dir)} direction"
-  end
-
-  defp format_reason({:invalid_pad_mode, pad, mode}) do
-    "Cannot invoke this action on pad #{inspect(pad)} as it works in #{inspect(mode)} mode"
-  end
-
-  defp format_reason({:invalid_demand_mode, pad, demand_mode}) do
-    "Cannot invoke this action on pad #{inspect(pad)} as it uses #{inspect(demand_mode)} demand mode"
-  end
-
-  defp format_reason({:eos_sent, pad}) do
-    "End of Stream has already been sent on pad #{inspect(pad)}"
-  end
-
-  defp format_reason({:invalid_caps, caps, accepted_caps}) do
+  defp format_reason({:unknown_action, doc_module}) do
     """
-    Trying to send caps that do not match the specification
-    Caps being sent: #{inspect(caps, pretty: true)}
-    Allowed caps spec: #{inspect(accepted_caps, pretty: true)}\
+    We looked everywhere, but couldn't find out what this action is supposed to do.
+    Make sure it's correct and valid for the component, callback, playback state or
+    other possible circumstances. See the docs for #{inspect(doc_module)} to check
+    which actions are supported and when you can return them.
     """
   end
+end
 
-  defp format_reason(:negative_demand) do
-    "Requested demand must be positive"
-  end
+defmodule Membrane.LinkError do
+  defexception [:message]
+end
 
-  defp format_reason({:invalid_event, event}) do
-    "Invalid event: #{inspect(event, pretty: true)}"
-  end
+defmodule Membrane.ElementError do
+  defexception [:message]
+end
 
-  defp format_reason(:caps_not_sent_before_first_buffer) do
-    "Tried to send a buffer, while caps have not been sent on this pad"
-  end
+defmodule Membrane.TimerError do
+  defexception [:message]
+end
 
-  defp format_reason(reason) do
-    "Unknown error: #{inspect(reason, pretty: true)}"
-  end
+defmodule Membrane.PadError do
+  defexception [:message]
 end
 
-defmodule Membrane.LinkError do
+defmodule Membrane.UnknownPadError do
   defexception [:message]
+
+  @impl true
+  def exception(opts) do
+    pad = Keyword.fetch!(opts, :pad)
+    module = Keyword.fetch!(opts, :module)
+    %__MODULE__{message: "Unknown pad #{inspect(pad)} of #{inspect(module)}"}
+  end
 end
 
-defmodule Membrane.ElementError do
+defmodule Membrane.InvalidPadDirectionError do
   defexception [:message]
+
+  @impl true
+  def exception(opts) do
+    opts = Map.new(opts)
+
+    %__MODULE__{
+      message: "Tried to send #{opts.what} via #{opts.direction} pad #{inspect(opts.pad)}"
```

Comment:
One thing is missing here - what is wrong with that send. I know it's included in the exception name, but I'd still include that in the message as well

```suggestion
      message: "Sending #{opts.what} via pad #{inspect(opts.pad)} with #{opts.direction} direction is not allowed"
```